### PR TITLE
make TransformerLayer accept a `bshd` or `sbhd` tensor format

### PR DESCRIPTION
**Pull request by sudhakarsingh27**
Changes:
1. `DotProductAttention` can interpret the input tensor in `sbhd` or `bshd` format but `MultiHeadAttention` and `TransformerLayer` aren't aware of this. This PR plumbs this information through `TransformerLayer` -> `MultiHeadAttention` -> `DotProductAttention`. 
2. `RotaryPositionEmbedding` also needs to be aware of this format to correctly apply the rope cache to the `q` and `k` inputs. 

Uses:
1. When replacing layers in HF models like `LlamaDecoderLayer` with `TransformerLayer`, it helps to have this control since HF is `bshd` by default and TE is `sbhd`.

### CODE REVIEW

**Review by cyanguwa — APPROVED** [review]
LGTM. Once the CI passes, can merge.

File: transformer_engine/pytorch/transformer.py [hunk 1]
```diff
@@ -158,6 +158,15 @@ class TransformerLayer(torch.nn.Module):
           The device on which the parameters of the model will allocated. It is the user's
           responsibility to ensure all parameters are moved to the GPU before running the
           forward pass.
+    hidden_states_format: str, default = 'sbhd'
```
Comment (ptrendx) [inline transformer_engine/pytorch/transformer.py:161]: Hmmm, `hidden_states_format` is very verbose for an option that will be used pretty frequently I think. Can we have something shorter here? I wonder if just `format` or maybe `input_format` would be enough`....

Also, `sbhd` is not really correct for a format to TransformerLayer, since it implies 4D tensors, while TransformerLayer actually has 3D tensors as input.

And yet another thing - since the options here are very limited, you can list them in the documentation in this line, similar to what was done in `normalization` option: `{ 'sbhd', 'bshd' }`

Comment (sudhakarsingh27) [reply transformer_engine/pytorch/transformer.py:161]: I think `input_format` (or maybe `input_layout`) sounds good. 
You're right about `sbhd` as it's only relevant for attention implementation. What if we make it `attn_input_format`? 

Otherwise, we'd either have to use [`sbd`, `bsd`] or [`batch_first`, `seq_first`] and translate them to `sbhd` or `bshd` inside the transformer layer. But then there will also be `td` format

File: transformer_engine/pytorch/attention.py [hunk 2]
```diff
@@ -845,11 +845,32 @@ def _rotate_half(x: torch.Tensor) -> torch.Tensor:
     return torch.cat((-x2, x1), dim=-1)
 
 
-def apply_rotary_pos_emb(t: torch.Tensor, freqs: torch.Tensor) -> torch.Tensor:
+def apply_rotary_pos_emb(t: torch.Tensor, freqs: torch.Tensor, tensor_format: str = "sbhd") -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (sudhakarsingh27) [reply transformer_engine/pytorch/attention.py:848]: I've added support for this test. Does it suffice?

File: transformer_engine/pytorch/transformer.py (hunk 1 above)
Comment (sudhakarsingh27) [reply transformer_engine/pytorch/transformer.py]: @ptrendx, what do you think?

File: transformer_engine/pytorch/attention.py [hunk 3]
```diff
@@ -846,18 +846,40 @@ def _rotate_half(x: torch.Tensor) -> torch.Tensor:
...
+        freqs = freqs.transpose(0,1) # [seq, 1, 1, dim] -> [1, seq, 1, dim]
+
     rot_dim = freqs.shape[-1]
     # ideally t_pass is empty so rotary pos embedding is applied to all tensor t
     t, t_pass = t[..., :rot_dim], t[..., rot_dim:]
 
     # first part is cosine component
     # second part is sine component, need to change signs with _rotate_half method
-    t = (t * freqs.cos()) + (_rotate_half(t) * freqs.sin())
+    t = (t * freqs.cos().to(t.dtype)) + (_rotate_half(t) * freqs.sin().to(t.dtype))
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention.py:882]: We can probably do one `.to(t.dtype)` here instead of two?

File: transformer_engine/pytorch/attention.py [hunk 4]
```diff
@@ -2579,6 +2601,16 @@ class MultiheadAttention(torch.nn.Module):
           The device on which the parameters of the model will allocated. It is the user's
           responsibility to ensure all parameters are moved to the GPU before running the
           forward pass.
+    qkv_format: str, default = `sbhd`
+            dimension format for `query_layer`, `key_layer` and `value_layer`,
+            {`sbhd`, `bshd`, `thd`}. `s` stands for the sequence length, `b` batch size,
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention.py:2606]: So with this PR, our `TransformerLayer` will handle `sbhd`, `bshd` formats, and `MultiHeadAttention` and `DotProductAttention` will handle `sbhd`, `bshd`, `thd` formats?

File: tests/pytorch/test_fused_attn.py [hunk 5]
```diff
@@ -666,10 +666,10 @@ def test_transformer_layer(dtype, model_configs, model, ckpt_attn, qkv_format, f
...
 @pytest.mark.parametrize("model", ["te_1_2", "te_2_0"])
-def test_te_layer_misc(dtype, model_configs, model):
+@pytest.mark.parametrize("qkv_format", ["bshd", "sbhd"])
+def test_te_layer_misc(dtype, model_configs, model, qkv_format):
     """Test TransformerLayer module with miscellanous settings"""
     ckpt_attn = True
-    qkv_format = "bshd"
     fused_qkv_params = True
     RoPE = True
     test_transformer_layer(dtype, model_configs, model,
```
Comment (cyanguwa) [inline tests/pytorch/test_fused_attn.py:675]: I think we need to remove the pytest parameter `qkv_format` on L592 because it's still running `test_transformer_layer` with `sbhd` instead of what's passed in? Might be worth double checking so we're actually testing both scenarios. Thanks!

File: tests/pytorch/test_fused_attn.py [hunk 6]
```diff
@@ -724,6 +724,10 @@ def _run_transformer_layer(
     # Create input tensor
     inp = torch.randn(config.max_seqlen_q, config.batch_size, config.hidden_size,
             dtype=dtype, device="cuda", requires_grad = True)
+    # In case the format to be tested is batch-first, need to transpose the
+    # input tensor.
+    if qkv_layout == "bshd":
```
Comment (cyanguwa) [inline tests/pytorch/test_fused_attn.py:729]: Maybe change `qkv_layout` to `qkv_format` here and on L708, since we've been using `qkv_format` in `test_transformer_layer` and `qkv_layout` actually means the more detailed layouts (like `sbhd_sbhd_sbhd`). Thanks!

**Review by cyanguwa — CHANGES_REQUESTED** [review]
Please fix a few small things. Looks good to me. Thanks!

File: transformer_engine/pytorch/attention.py (hunk 4 above)
Comment (sudhakarsingh27) [reply transformer_engine/pytorch/attention.py:2606]: Ah and that's a gap right? I think we could provide `TransformerLayer` with `thd` as well. The `TransformerLayer` just passes that argument to the `MultiHeadAttention` layer anyways without doing anything with it itself.

Comment (sudhakarsingh27) [reply transformer_engine/pytorch/attention.py:2606]: @cyanguwa should I add `thd` format to the docstring?

File: transformer_engine/pytorch/attention.py (hunk 3 above)
Comment (sudhakarsingh27) [reply transformer_engine/pytorch/attention.py]: Makes sense!

File: transformer_engine/pytorch/attention.py [hunk 7]
```diff
@@ -846,18 +846,40 @@ def _rotate_half(x: torch.Tensor) -> torch.Tensor:
...
     """
+    assert tensor_format in ("sbhd", "bshd"),("Only formats `sbhd` or `bshd` "
+                                              "are supported for input tensor "
+                                              "`t`.")
+    max_seq_len = freqs.shape[0]
+    cur_seq_len = t.shape[1] if tensor_format == "bshd" else t.shape[0]
+
+    # Only apply the rotary embeddings up to the sequence length of the running
+    # input.
+    assert cur_seq_len <= max_seq_len
```
[the remark this answers is not in the document]
Comment (sudhakarsingh27) [reply transformer_engine/pytorch/attention.py]: Sounds good!

[the remark this answers is not in the document]
Comment (sudhakarsingh27) [reply transformer_engine/pytorch/attention.py]: done!

File: transformer_engine/pytorch/attention.py (hunk 3 above)
Comment (sudhakarsingh27) [reply transformer_engine/pytorch/attention.py]: yes good catch!

**Comment by ptrendx** [comment]
@sudhakarsingh27 Please resolve the merge conflicts

@cyanguwa Could you rereview?

File: transformer_engine/pytorch/attention.py (hunk 4 above)
Comment (cyanguwa) [reply transformer_engine/pytorch/attention.py:2596]: I guess you shouldn't because TransformerLayer doesn't support `thd`. I think it's ok to keep `qkv_format = {sbhd, bshd}` for TransformerLayer and MultiHeadAttention, and `{sbhd, bshd, thd}` for DPA.

Comment (sudhakarsingh27) [reply transformer_engine/pytorch/attention.py]: sounds good!

**Comment by sudhakarsingh27** [comment]
/te-ci/pytorch

File: transformer_engine/pytorch/attention.py [hunk 8]
```diff
@@ -2821,6 +2844,16 @@ class MultiheadAttention(torch.nn.Module):
...
           responsibility to ensure all parameters are moved to the GPU before running the
           forward pass.
+    qkv_format: str, default = `sbhd`
+            dimension format for `query_layer`, `key_layer` and `value_layer`,
+            {`sbhd`, `bshd`}. `s` stands for the sequence length, `b` batch size,
+            `h` the number of heads, `d` head size, and `t` the total number of sequences
+            in a batch, with `t = sum(s_i), for i = 0...b-1`. `sbhd` and `bshd` formats
+            are used for when sequences in a batch are of equal length or padded to
+            equal length, and the `thd` format is used for when sequences in a batch
+            have different lengths. Please note that these formats do not reflect how
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention.py:2854]: You might want to remove the `thd` related explanations since we don't support thd here anyway? Thanks.

File: transformer_engine/pytorch/transformer.py [hunk 9]
```diff
@@ -168,6 +168,14 @@ class TransformerLayer(torch.nn.Module):
...
           responsibility to ensure all parameters are moved to the GPU before running the
           forward pass.
+    attn_input_format: {'sbhd', 'bshd'}, default = 'sbhd'
+                         This controls whether the dimensions of the
+                         intermediate hidden states is 'batch first' ('bshd') or
+                         'sequence first' ('sbhd'). `s` stands for the sequence
+                         length, `b` batch size, `h` the number of heads, `d`
+                         head size. Note that these formats are very closely
+                         related to the `qkv_format` in the `MultiHeadAttention`
+                         and `DotProductAttention` modules.
```
Comment (cyanguwa) [inline transformer_engine/pytorch/transformer.py:178]: Do we want to call the parameter `attn_input_format` here and then `qkv_format` in MHA and DPA?

Comment (sudhakarsingh27) [reply transformer_engine/pytorch/transformer.py:178]: Yes, I think @ptrendx is also okay with it.

**Review by cyanguwa — APPROVED** [review]
LGTM

**Comment by cyanguwa** [comment]
@sudhakarsingh27 Please fix CI before merging -- looks like there's a few failed jobs.
